File: src/degenbot/balancer/libraries/weighted_math.py
from degenbot.balancer.libraries.constants import ONE
from degenbot.balancer.libraries.fixed_point import (
    complement,
    div_down,
    div_up,
    mul_down,
    mul_up,
    pow_down,
    pow_up,
)
from degenbot.exceptions import EVMRevertError

_MIN_WEIGHT = int(0.01 * 10**18)
# Having a minimum normalized weight imposes a limit on the maximum number of tokens;
# i.e., the largest possible pool is one where all tokens have exactly the minimum weight.
_MAX_WEIGHTED_TOKENS = 100

# Pool limits that arise from limitations in the fixed point power function (and the imposed 1:100
# maximum weight ratio).

# Swap limits: amounts swapped may not be larger than this percentage of total balance.
_MAX_IN_RATIO = int(0.3 * 10**18)
_MAX_OUT_RATIO = int(0.3 * 10**18)


# Invariant growth limit: non-proportional joins cannot cause the invariant to increase by more than
# this ratio.
_MAX_INVARIANT_RATIO = 3 * 10**18
# Invariant shrink limit: non-proportional exits cannot cause the invariant to decrease by less than
# this ratio.
_MIN_INVARIANT_RATIO = int(0.7 * 10**18)


def calculate_invariant(normalized_weights: list[int], balances: list[int]) -> int:
    invariant = ONE

    for i in range(len(normalized_weights)):
        invariant = mul_down(invariant, pow_down(balances[i], normalized_weights[i]))

    if invariant <= 0:
        raise EVMRevertError(error="ZERO_INVARIANT")

    return invariant


def _calc_out_given_in(
    balance_in: int,
    weight_in: int,
    balance_out: int,
    weight_out: int,
    amount_in: int,
) -> int:
    """
    Computes how many tokens can be taken out of a pool if `amountIn` are sent, given the
    current balances and weights.
    """

    # ********************************************************************************************
    # outGivenIn                                                                                //
    # aO = amountOut                                                                            //
    # bO = balanceOut                                                                           //
    # bI = balanceIn              /      /            bI             \    (wI / wO) \           //
    # aI = amountIn    aO = bO * |  1 - | --------------------------  | ^            |          //
    # wI = weightIn               \      \       ( bI + aI )         /              /           //
    # wO = weightOut                                                                            //
    # *******************************************************************************************/

    # Amount out, so we round down overall.

    # The multiplication rounds down, and the subtrahend (power) rounds up (so the base rounds up
    # too). Because bI / (bI + aI) <= 1, the exponent rounds down.

    # // Cannot exceed maximum in ratio
    if amount_in > mul_down(
        balance_in,
        _MAX_IN_RATIO,
    ):
        raise EVMRevertError(error="MAX_IN_RATIO")

    denominator = balance_in + amount_in
    base = div_up(balance_in, denominator)
    exponent = div_down(weight_in, weight_out)
    power = pow_up(base, exponent)
    return int(mul_down(balance_out, complement(power)))


def _calc_in_given_out(
    balance_in: int,
    weight_in: int,
    balance_out: int,
    weight_out: int,
    amount_out: int,
) -> int:
    """
    Computes how many tokens must be sent to a pool in order to take `amountOut`, given the
    current balances and weights.
    """

    # ********************************************************************************************
    # inGivenOut                                                                                //
    # aO = amountOut                                                                            //
    # bO = balanceOut                                                                           //
    # bI = balanceIn              /  /            bO             \    (wO / wI)      \          //
    # aI = amountIn    aI = bI * |  | --------------------------  | ^            - 1  |         //
    # wI = weightIn               \  \       ( bO - aO )         /                   /          //
    # wO = weightOut                                                                            //
    # *******************************************************************************************/

    # Amount in, so we round up overall.

    # The multiplication rounds up, and the power rounds up (so the base rounds up too).
    # Because b0 / (b0 - a0) >= 1, the exponent rounds up.

    _balance_in = balance_in
    _weight_in = weight_in
    _balance_out = balance_out
    _weight_out = weight_out
    _amount_out = amount_out

    # Cannot exceed maximum out ratio
    if not (_amount_out <= mul_down(_balance_out, _MAX_OUT_RATIO)):
        raise EVMRevertError(error="MAX_OUT_RATIO")

    _base = div_up(_balance_out, _balance_out - _amount_out)
    _exponent = div_up(_weight_out, _weight_in)
    _power = pow_up(_base, _exponent)
    # Because the base is larger than one (and the power rounds up), the power should always be
    # larger than one, so the following subtraction should never revert.
    _ratio = _power - ONE
    return mul_up(_balance_in, _ratio)


def _subtract_swap_fee_amount(amount: int, fee_percentage: int) -> int:
    """
    Subtracts swap fee amount from `amount`, returning a lower value.
    """

    # This returns amount - fee amount, so we round up (favoring a higher fee amount).
    fee_amount = mul_up(amount, fee_percentage)
    return amount - fee_amount

File: rust/crates/degenbot-balancer-math/src/weighted_math.rs
//! Balancer V2 `WeightedMath` — invariant + swap math + fee helpers.
//! Direct port of `src/degenbot/balancer/libraries/weighted_math.py`.
//!
//! All quantities are non-negative `U256` (balances, weights, amounts, fee
//! percentages). Overflow checks delegate to [`crate::fixed_point`] (matching
//! Python's `MAX_UINT256`-checked Solidity `SafeMath` reverts).

use alloy::primitives::U256;

use crate::constants::{PowVersion, ONE};
use crate::fixed_point::{
    complement, div_down, div_up, mul_down, mul_up, pow_down, pow_up, Result,
};

/// `_MIN_WEIGHT = 0.01e18` — minimum allowed normalized weight.
#[allow(dead_code)]
const MIN_WEIGHT: U256 = U256::from_limbs([10_000_000_000_000_000, 0, 0, 0]);

/// `_MAX_WEIGHTED_TOKENS = 100` — the largest possible pool given the
/// minimum normalized weight.
#[allow(dead_code)]
const MAX_WEIGHTED_TOKENS: u64 = 100;

/// `_MAX_IN_RATIO = 0.3e18` — swap inputs may not exceed 30% of the
/// token's balance.
const MAX_IN_RATIO: U256 = U256::from_limbs([300_000_000_000_000_000, 0, 0, 0]);

/// `_MAX_OUT_RATIO = 0.3e18` — swap outputs may not exceed 30% of the
/// token's balance.
const MAX_OUT_RATIO: U256 = U256::from_limbs([300_000_000_000_000_000, 0, 0, 0]);

// Invariant growth / shrink limits (unused by swap math; ported for fidelity).
#[allow(dead_code)]
const MAX_INVARIANT_RATIO: U256 = U256::from_limbs([3_000_000_000_000_000_000, 0, 0, 0]);
#[allow(dead_code)]
const MIN_INVARIANT_RATIO: U256 = U256::from_limbs([700_000_000_000_000_000, 0, 0, 0]);

/// Calculate the weighted-product invariant
/// `invariant = Π(balance_i ^ weight_i)`.
///
/// `version` selects the `pow_down` implementation (V1 general / V2 fast
/// paths) matching the pool's deployed `FixedPoint` library.
pub fn calculate_invariant(
    normalized_weights: &[U256],
    balances: &[U256],
    version: PowVersion,
) -> Result<U256> {
    let mut invariant = ONE;
    for i in 0..normalized_weights.len() {
        invariant = mul_down(
            invariant,
            pow_down(balances[i], normalized_weights[i], version)?,
        )?;
    }
    if invariant.is_zero() {
        return Err(crate::BalancerMathError::ZeroInvariant);
    }
    Ok(invariant)
}

/// `_calc_out_given_in(balance_in, weight_in, balance_out, weight_out, amount_in, version)`.
///
/// Compute how many tokens can be taken out of a pool if `amount_in` are
/// sent, given the current balances and weights. Rounds down overall
/// (amount out → favor the pool).
///
/// ```text
/// aO = bO * (1 - (bI / (bI + aI)) ^ (wI / wO))
/// ```
pub fn calc_out_given_in(
    balance_in: U256,
    weight_in: U256,
    balance_out: U256,
    weight_out: U256,
    amount_in: U256,
    version: PowVersion,
) -> Result<U256> {
    // Cannot exceed maximum in ratio (30% of balance_in).
    if amount_in > mul_down(balance_in, MAX_IN_RATIO)? {
        return Err(crate::BalancerMathError::MaxInRatio);
    }
    let denominator = balance_in
        .checked_add(amount_in)
        .ok_or(crate::BalancerMathError::AddOverflow)?;
    let base = div_up(balance_in, denominator)?;
    let exponent = div_down(weight_in, weight_out)?;
    let power = pow_up(base, exponent, version)?;
    mul_down(balance_out, complement(power))
}

/// `_calc_in_given_out(balance_in, weight_in, balance_out, weight_out, amount_out, version)`.
///
/// Compute how many tokens must be sent in to take `amount_out`. Rounds up
/// overall (amount in → favor the pool).
///
/// ```text
/// aI = bI * ((bO / (bO - aO)) ^ (wO / wI) - 1)
/// ```
pub fn calc_in_given_out(
    balance_in: U256,
    weight_in: U256,
    balance_out: U256,
    weight_out: U256,
    amount_out: U256,
    version: PowVersion,
) -> Result<U256> {
    // Cannot exceed maximum out ratio (30% of balance_out).
    if amount_out > mul_down(balance_out, MAX_OUT_RATIO)? {
        return Err(crate::BalancerMathError::MaxOutRatio);
    }
    let denominator = balance_out
        .checked_sub(amount_out)
        .ok_or(crate::BalancerMathError::SubOverflow)?;
    let base = div_up(balance_out, denominator)?;
    let exponent = div_up(weight_out, weight_in)?;
    let power = pow_up(base, exponent, version)?;
    // The base is >= 1 (and power rounds up), so `power >= ONE`; the
    // subtraction cannot underflow.
    let ratio = power
        .checked_sub(ONE)
        .ok_or(crate::BalancerMathError::SubOverflow)?;
    mul_up(balance_in, ratio)
}

/// `_subtract_swap_fee_amount(amount, fee_percentage)`.
///
/// Subtract the swap fee from `amount`: `amount - mul_up(amount, fee)`.
/// Rounds the fee up (favoring a higher fee deduction).
pub fn subtract_swap_fee_amount(amount: U256, fee_percentage: U256) -> Result<U256> {
    let fee_amount = mul_up(amount, fee_percentage)?;
    amount
        .checked_sub(fee_amount)
        .ok_or(crate::BalancerMathError::SubOverflow)
}

/// `_add_swap_fee_amount(amount, fee_percentage)`.
///
/// Add the swap fee on top of `amount`: `div_up(amount, ONE - fee)`.
/// Rounds the amount up (favoring a higher fee addition).
pub fn add_swap_fee_amount(amount: U256, fee_percentage: U256) -> Result<U256> {
    div_up(amount, complement(fee_percentage))
}
